compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -139,7 +140,32 @@ impl<'thir, 'tcx> Visitor<'thir, 'tcx> for UnsafetyVisitor<'tcx> {
         }
     }
 
+    fn visit_pat(&mut self, pat: &Pat<'tcx>) {
+        use pattern::PatKind::*;
+
+        match *pat.kind {
+            AscribeUserType { .. } | Constant { .. } => {}
+            _ => {
+                if let ty::Adt(adt_def, _) = pat.ty.kind() {
+                    // binding to wildcard is okay since that never reads anything and stops double errors
+                    // with implict wildcard branches from `if let`s
+                    if adt_def.is_union() && !matches!(*pat.kind, Binding { .. } | Wild) {
```

Comment:
The comment explains the `Wild`, but why is binding to `Binding` okay? (I have no idea what that even means, THIR is foreign to me.)

Reply:
I've added some comments to better explain this. If the pattern kind is `Binding` then the union itself is being looked at, not a union field. It is safe to assign the whole union to another variable if you don't look at the fields:
```rust
let x = Foo { bar: 5 };
let y = x; // safe
```